crates/walrus-sui/src/client/retry_client/download_handler.rs

```diff
@@ -0,0 +1,606 @@
+            last_tracked_failed_checkpoint: Arc::new(AtomicU64::new(0)),
+            current_consecutive_failures_for_tracked_checkpoint: Arc::new(AtomicUsize::new(0)),
+        }
+    }
+
+    /// Called on primary success for a specific sequence_number
+    pub fn record_primary_success(&self, sequence_number: u64) {
+        self.last_success.store(Instant::now(), Ordering::Relaxed);
+        self.num_failures.store(0, Ordering::Relaxed);
+
+        // Reset consecutive failure tracking if this was the specific checkpoint being tracked
+        if self.last_tracked_failed_checkpoint.load(Ordering::Relaxed) == sequence_number {
+            self.last_tracked_failed_checkpoint
+                .store(0, Ordering::Relaxed);
+            self.current_consecutive_failures_for_tracked_checkpoint
+                .store(0, Ordering::Relaxed);
+            tracing::debug!(
+                seq = sequence_number,
+                "Reset consecutive failure tracking after primary success."
+            );
+        }
+    }
+
+    /// Called on fallback success for a specific sequence_number
+    pub fn record_fallback_success_for_sequence(&self, sequence_number: u64) {
+        if self.last_tracked_failed_checkpoint.load(Ordering::Relaxed) == sequence_number {
+            self.last_tracked_failed_checkpoint
+                .store(0, Ordering::Relaxed);
+            self.current_consecutive_failures_for_tracked_checkpoint
+                .store(0, Ordering::Relaxed);
+            tracing::debug!(
+                seq = sequence_number,
+                "Reset consecutive failure tracking after fallback success."
+            );
+        }
+    }
+
+    /// Determines the initial action for fetching a checkpoint
+    pub fn initial_download_decision(&self) -> CheckpointDownloadDecision {
+        if self.config.is_none() {
+            return CheckpointDownloadDecision::AttemptPrimary;
+        }
+        if Instant::now() >= self.skip_primary_until.load(Ordering::Relaxed) {
+            CheckpointDownloadDecision::AttemptPrimary
+        } else {
+            tracing::debug!(
+                "Skipping primary RPC (overall policy) until {:?}, attempting fallback directly.",
+                self.skip_primary_until.load(Ordering::Relaxed)
+            );
+            CheckpointDownloadDecision::TryFallback(None)
+        }
+    }
+
+    /// Updates all relevant failure counters based on the current failure
+    /// and returns a snapshot of the counts relevant to the decision.
+    /// Prioritizes tracking failures for smaller sequence numbers.
+    pub fn update_and_get_failure_stats(&self, sequence_number: u64) -> CheckpointFailureStats {
+        let overall_failures_count = self.num_failures.fetch_add(1, Ordering::Relaxed) + 1;
+
+        let (consecutive_failures_for_this_seq, is_this_sequence_now_tracked) = loop {
+            let current_tracked_seq = self.last_tracked_failed_checkpoint.load(Ordering::Acquire);
```

Comment:
Question: multiple orderings are used in this function, and how do you decide which level to use in each operation?

Reply:
I think this was to ensure that updates made to `last_tracked_failed_checkpoint` by other threads during CAS are visible during this load, otherwise with `Ordering::Relaxed` I don't think we get any such guarantee